node/router/src/lib.rs

```diff
@@ -251,6 +256,53 @@ impl<N: Network> Router<N> {
         !self.is_local_ip(ip) && !is_bogon_ip(ip.ip()) && !is_unspecified_or_broadcast_ip(ip.ip())
     }
 
+    /// Returns `true` if the message version is valid.
+    pub fn is_valid_message_version(&self, message_version: u32) -> bool {
+        // Fetch the latest message version.
+        let latest_message_version = Message::<N>::latest_message_version();
+
+        // Determine the minimum message version this node will accept, based on its role.
+        // - Provers always operate at the latest message version.
+        // - Validators and clients may accept older versions, depending on their current block height.
+        //
+        // Example scenario:
+        // - At block height `X`, the protocol upgrades to message version from `Y-1` to `Y`.
+        // - Client A upgrades and starts using message version `Y`.
+        // - Client B has not upgraded and still uses message version `Y-1`.
+        // - Until block `X`, they stay connected and can communicate.
+        // - After block `X`, Client A will reject messages from Client B.
+        let lowest_accepted_message_version = match self.node_type {
+            // Provers should always use the latest version.
+            NodeType::Prover => Message::<N>::latest_message_version(),
+            // Validators and clients accept messages from lower version based on the migration height.
+            NodeType::Validator | NodeType::Client => {
+                let latest_block_height = self.ledger.latest_block_height();
+                let versions = Message::<N>::VERSIONS;
+                // TODO (raychu86): Refactor `consensus_config_value` to support generic constant types.
```

Comment:
It would be great if Ray or Kai can take one more 30 minute stab at this before we merge the PR, if its too cumbersome we can leave it.

Reply:
Note logic slightly changed in [3eca0df](https://github.com/ProvableHQ/snarkOS/pull/3596/commits/3eca0dfa4df035e58f66ebc482c3bcc94dd3119a) which deviates from the macro.